programs/uploader/src/processor.rs

```diff
@@ -0,0 +1,176 @@
+use {
+    crate::{
+        instruction::{parse_program_instruction, ProgramInstruction},
+        state,
+    },
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        clock::Clock,
+        entrypoint,
+        entrypoint::ProgramResult,
+        msg,
+        program::invoke_signed,
+        program_error::ProgramError,
+        pubkey::Pubkey,
+        rent::Rent,
+        system_instruction,
+        sysvar::Sysvar,
+    },
+};
+
+// Declare and export the program's entrypoint.
+entrypoint!(process_instruction);
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    match parse_program_instruction(instruction_data)? {
+        ProgramInstruction::CreateBucket {
+            data,
+            size,
+            bump_seed,
+        } => Processor::create_bucket(program_id, accounts, data, size as usize, bump_seed),
+        ProgramInstruction::PutIntoBucket { data, offset } => {
+            Processor::put_into_bucket(program_id, accounts, data, offset as usize)
+        }
+    }
+}
+
+fn adler32(data: &[u8]) -> u32 {
+    let mut hash = adler32::RollingAdler32::new();
+    hash.update_buffer(data);
+    hash.hash()
+}
+
+pub struct Processor;
+impl Processor {
+    fn create_bucket(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        data: &[u8],
+        size: usize,
+        bump_seed: u8,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+
+        let authority_account = next_account_info(account_info_iter)?;
+        let payer_account = next_account_info(account_info_iter)?;
+        let data_bucket_account = next_account_info(account_info_iter)?;
+        let system_program_account = next_account_info(account_info_iter)?;
+
+        let authority_key = *authority_account.signer_key().ok_or_else(|| {
+            msg!("Authority account must be a signer");
+            ProgramError::MissingRequiredSignature
+        })?;
+
+        let payer_key = *payer_account.signer_key().ok_or_else(|| {
+            msg!("Payer account must be a signer");
+            ProgramError::MissingRequiredSignature
+        })?;
+
+        // Use a derived address to ensure that an address table can never be
+        // initialized more than once at the same address.
+        let adler_csum = adler32(data);
+        let adler_csum = adler_csum.to_be_bytes();
+        let adler_csum = adler_csum.as_slice();
```

Comment:
The Data Bucket account's seed contains now also the Adler32 checksum of the first chunk, to distinct the different transactions from each others :)